### PR TITLE
fix(#279): strategy deps are shaded and excluded their transitive deps

### LINKED ISSUE #279: Be as less intrusive as possible with the dependencies
##### Issue Overview

Most of the strategy implementations, core and also surefire provider has several transitive dependencies. These transitive deps may cause conflicts, classloading issues or other problems such as failures when transitive dependencies are forbidden (using force plugin).
All dependencies that are added to the pom should be as less intrusive as possible.
Proposed solution:
* surefire provider & core should be handled by this issue: #277 
* all strategies should be shaded as well
* when the dependency is added to the effective `pom.xml` file, an exclusion for all dependencies should be set
* jar should be minimized

##### Steps To Reproduce
1. clone https://github.com/wildfly/wildfly
2. install ST
3. run build with some strategy specified

### PR DESCRIPTION
**Pull request by MatousJobanek**
#### Short description of what this resolves:

All strategies are distributed as shaded jars and when they are added to the model, all their transitive dependencies are excluded.

#### Changes proposed in this pull request:

- added shaded plugin configuration to all strategies
- added complete exclusion of transitive deps for all default (shaded) strategy dependencies
- introduced whole coordinate format: `groupId:artifactId:packaging:classifier:version:scope`

- [x]  not added - relocations: https://github.com/arquillian/smart-testing/issues/279#issuecomment-350393260

Fixes #279

### CODE REVIEW

File: mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/dependencies/StrategyDependencyResolver.java [hunk 1]
```diff
@@ -67,26 +67,46 @@ private Properties loadDefaultMapping() {
...
+            dependency.setClassifier(coordinates[3]);
+            dependency.setVersion(coordinates[4]);
+        }
+        if (number == 6) {
+            dependency.setScope(coordinates[5]);
+        }
+        if (dependency.getVersion() == null || dependency.getVersion().isEmpty()) {
+            dependency.setVersion(ExtensionVersion.version().toString());
+        }
+        if (excludeTrainsitive) {
```
Comment (bartoszmajsak) [inline mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/dependencies/StrategyDependencyResolver.java]: Typo - should be `transitive`, right? Not only in this scope, but everywhere in this class :)

File: mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/dependencies/StrategyDependencyResolver.java [hunk 2]
```diff
@@ -67,26 +67,46 @@ private Properties loadDefaultMapping() {
...
-                        version = gav[2];
-                    }
-                    dependency.setVersion(version);
-                    dependency.setScope("runtime");
-                    return dependency;
-                }));
+                key -> createDependency((String) properties.get(key), excludeTrainsitive)));
+    }
+
+    private Dependency createDependency(String coordinatesString, boolean excludeTrainsitive) {
```
Comment (bartoszmajsak) [inline mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/dependencies/StrategyDependencyResolver.java]: Maybe we can move it to external class called `MavenCoordinatesResolver` and also provide unit tests for just this feature?

File: mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/dependencies/StrategyDependencyResolver.java [hunk 3]
```diff
@@ -67,26 +67,46 @@ private Properties loadDefaultMapping() {
...
+        final String[] coordinates = coordinatesString.split(":");
+        int number = coordinates.length;
```
Comment (bartoszmajsak) [inline mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/dependencies/StrategyDependencyResolver.java]: `number` is quite generic, maybe `amountOfCoordinates` or even just inline.

File: mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/dependencies/StrategyDependencyResolver.java (hunk 2 above)
Comment (MatousJobanek) [reply mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/dependencies/StrategyDependencyResolver.java]: I had the same feeling when I was writing it. But I was too lazy :-D

Comment (MatousJobanek) [reply mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/dependencies/StrategyDependencyResolver.java]: Ok, I moved it to separated class and also created a test with custom assertion for it

**Pull request merged**